### PR TITLE
[SPARK-15992] [MESOS] Refactor MesosCoarseGrainedSchedulerBackend offer consideration

### PR DESCRIPTION
**Pull request by drcrallen**
The offer acceptance workflow is a little hard to follow and not very extensible for future considerations for offers. This is a patch that makes the workflow a little more explicit in its handling of offer resources.

### CODE REVIEW

**Comment by rxin** [comment]
Thanks for the pull request, but the refactoring seems to have made it more difficult to read?

**Comment by drcrallen** [comment]
I broke something in the test suite, evaluating and I'll open this back in a bit

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -382,59 +382,97 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
...
+   * @param sc Spark context
+   * @param availableCpus The available CPUs
+   * @param availableMem The available Memory
+   * @return Tuple of CPU (integer cores)  and Memory (integer MB) desired
+   */
+  private[mesos]
+  def calculateUsableResources(sc: SparkContext, availableCpus: Int, availableMem: Int):
+  Option[(Int, Int)] = {
+    val desiredMemory = executorMemory(sc)
+    val desiredCpu = executorCores(availableCpus)
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: I like the idea, but it does make it harder to read. Can we refactor so we can collapse global and resource limits, and introduce like a case class for each condition, where it knows how to verify and logs error message when it doesn't match? 
I think it makes adding and understanding how offers are matched better.

**Comment by tnachen** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala (hunk 1 above)
Comment (drcrallen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: I can do that. Will be a few weeks before I can tackle this again though.

**Comment by HyukjinKwon** [comment]
(ping @drcrallen)
